sql/item_jsonfunc.cc

```diff
@@ -1150,6 +1150,76 @@ static int check_contains(json_engine_t *js, json_engine_t *value)
 }
 
 
+longlong Item_func_json_equals::val_int()
```

Comment:
Great job on the first try here! A few meta comments first. Let's fix these, then we'll dig into the small details such as minor coding style errors etc.

1. You've taken inspiration from Item_func_json_contains, however the variable names don't quite match what we should have here. `je1` and `je2` for engines is enough. Either use `val1` and `val2` without `js1` or `js2` or the other way around.
2. There isn't actually a need for 4 `json_engine_t`'s. You can restart the parser by doing a json_scan_start.
3. It looks like there is a bug within check_contains, or its semantics are not 100% clear, due to that duplicates in an array test case that you've encountered. I will let you investigate that as it's a good exercise in debugging. If you get stuck, ping me.

Ultimately, I don't want us to subclass Item_func_json_contains in the end. Now that you've got this example working, can you try to just inherit from Item_bool_func and see which bits of code you actually need?

Notice the a2_parsed and a2_constant flags are used to "speed up" a single call and that is the "val_json" call.
The idea behind that (for Item_func_json_contains) is: If an argument to a function never changes during execution (`is_const()`), there is no need to call val_json over and over, for every row in the result set. It should be fetched once and "cached".

Consider that now it's also possible to optimize this further: do you really need to check for equality more than once if *both* arguments are constant? :)